client/cache/read_through.go

```diff
@@ -0,0 +1,67 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"github.com/beego/beego/v2/core/berror"
+	"strings"
+	"time"
+)
+
+// readThroughCache is a decorator
+// add the read through function to the original Cache function
+type readThroughCache struct {
+	Cache
+	Expiration time.Duration
+	LoadFunc   func(ctx context.Context, key string) (any, error)
+}
+
+// NewReadThroughCache create readThroughCache
+func NewReadThroughCache(cache Cache, expiration time.Duration,
+	loadFunc func(ctx context.Context, key string) (any, error)) (Cache, error) {
+	if loadFunc == nil {
+		return nil, berror.Error(InvalidLoadFunc, "loadFunc cannot be nil")
+	}
+	return &readThroughCache{
+		Cache:      cache,
+		Expiration: expiration,
+		LoadFunc:   loadFunc,
+	}, nil
+}
+
+// Get cache from readThroughCache
+func (c *readThroughCache) Get(ctx context.Context, key string) (any, error) {
+	val, err := c.Cache.Get(ctx, key)
+	if val == nil || err != nil {
+		val, err = c.LoadFunc(ctx, key)
+		if err != nil {
+			return nil, berror.Wrap(
+				err, KeyNotExist, "cache unable to load data")
```

Comment:
Define a new error instead of using KeyNotExit, i.e FailedToLoadData. As you can see, the key may be exist.

Reply:
The error code shout start with 4 because it's user's input error(the load func is invalid)